### PR TITLE
Demonstrate type in DataObject::MakeOutput

### PR DESCRIPTION
**Pull request by blowekamp**
This but was encountered when setting an output to nullptr.

### CODE REVIEW

File: Modules/Core/Common/src/itkProcessObject.cxx [hunk 1]
```diff
@@ -90,6 +90,7 @@ ::MakeOutput( const DataObjectIdentifierType & name )
     {
     return this->MakeOutput( this->MakeIndexFromOutputName(name) );
     }
+  // DataObject::New() returns is from Object, and returns only Object!
   return static_cast<DataObject *>(DataObject::New().GetPointer());
```
Comment (blowekamp) [inline Modules/Core/Common/src/itkProcessObject.cxx]: Here an invalid upcast occours on a concrete `itk::Object`.

**Comment by blowekamp** [comment]
I'm not sure what to do about this bug I encounter.

File: Modules/Core/Common/src/itkProcessObject.cxx [hunk 2]
```diff
@@ -90,6 +90,7 @@ ::MakeOutput( const DataObjectIdentifierType & name )
     {
     return this->MakeOutput( this->MakeIndexFromOutputName(name) );
     }
+  // DataObject::New() returns is from Object, and returns only Object!
```
Comment (dzenanz) [inline Modules/Core/Common/src/itkProcessObject.cxx]: itkDataObject.h is missing `itkTypeMacro`.

Comment (dzenanz) [reply Modules/Core/Common/src/itkProcessObject.cxx]: Oops, it has that but it is missing definition of new.

Comment (dzenanz) [reply Modules/Core/Common/src/itkProcessObject.cxx]: `Object` has manually defined new, whereas `DataObject` has neither `itkNewMacro` nor manually defined/overridden `New`.

Comment (dzenanz) [reply Modules/Core/Common/src/itkProcessObject.cxx]: Perhaps both `Object` and `DataObject` should use `itkNewMacro`?

Comment (blowekamp) [reply Modules/Core/Common/src/itkProcessObject.cxx]: Yup. The questions is: Should` DataObject` be instantiable or an abstract base class?

Comment (dzenanz) [reply Modules/Core/Common/src/itkProcessObject.cxx]: Is `DataObject` currently instantiable? It looks like it is. So just fix the way it is instantiated? If I am wrong, what are the abstract members?

Comment (blowekamp) [reply Modules/Core/Common/src/itkProcessObject.cxx]: New method for `DataObject` seems to work just fine. Re: `Object::New` likely has to be in the cxx file due to a circular dependency with using the `ObjectFactory`

**Pull request merged**